### PR TITLE
DOC: Improve documentation for fill_diagonal

### LINKED ISSUE #25077: DOC: imprecise definition in numpy.fill_diagonal
### Issue with current documentation:

> `the diagonal is the list of locations with indices a[i, ..., i] all identical. `

This does not read correctly.  `a[i, ..., i] `is the indexed array, not a list of locations.  Maybe it should be:

`the diagonal is the list of locations a[i, ..., i] with indices i, ..., i all identical. `

?

### Idea or request for content:

_No response_

### PR DESCRIPTION
**Pull request by linus-md**
Closes #25077.

I wrote a detailed description of what values the diagonal comprises.
I hope this provides value.

### CODE REVIEW

**Comment by linus-md** [comment]
I do not understand if this "Invoke-WebRequest : Unable to connect to the remote server" has anything to do with my changes.

File: numpy/lib/_index_tricks_impl.py [hunk 1]
```diff
@@ -788,9 +788,12 @@ def _fill_diagonal_dispatcher(a, val, wrap=None):
...
     """Fill the main diagonal of the given array of any dimensionality.
 
-    For an array `a` with ``a.ndim >= 2``, the diagonal is the list of
-    locations with indices ``a[i, ..., i]`` all identical. This function
-    modifies the input array in-place, it does not return a value.
+    For an array `a` with ``a.ndim >= 2`` when ``wrap == False`` 
+    the diagonal consists of the values with indices 
+    ``[(i, ..., i) for i in range(min(a.shape))]``,
+    otherwise the diagonal consists of values with indices
+    ``[(i % a.shape[0], ..., i % a.shape[-1]) for i in range(max(a.shape))]``.
```
Comment (rossbar) [inline numpy/lib/_index_tricks_impl.py:795]: IMO these formulae are confusing (and I think incorrect in general). I'd prefer something along the lines of [@mdhaber 's suggestion](https://github.com/numpy/numpy/compare/main...mdhaber:numpy:gh25077) which I think is sufficient to address the ambiguity raised in the original issue but preserves the brevity and clarity of the short description.

Comment (linus-md) [reply numpy/lib/_index_tricks_impl.py]: Ok, I made some changes. But should this PR exist in light of [@mdhaber 's suggestion](https://github.com/numpy/numpy/compare/main...mdhaber:numpy:gh25077), which I did not see before?

**Comment by ngoldbaum** [comment]
It doesn't matter who makes the PR, as long as it gets made :)

**Comment by linus-md** [comment]
Great @ngoldbaum! Do you have any other feedback? :)

File: numpy/lib/_index_tricks_impl.py [hunk 2]
```diff
@@ -789,14 +789,14 @@ def fill_diagonal(a, val, wrap=False):
...
+    values ``a[i, ..., i]`` with indices `i` all identical.  This function 
+    modifies the input array in-place without returning a value.
 
     Parameters
     ----------
     a : array, at least 2-D.
-      Array whose diagonal is to be filled, it gets modified in-place.
-
+      Array whose diagonal is to be filled in-place.
+
```
Comment (ngoldbaum) [inline numpy/lib/_index_tricks_impl.py:799]:
```suggestion

```

Just the trailing whitespace here, otherwise this looks good to me.

**Review by rossbar — APPROVED** [review]
Thanks @linus-md , LGTM! I pushed up one last minor rst formatting nit and added @mdhaber as a co-author since we've borrowed heavily from his wording suggestion. Thanks all!

**Pull request merged**